### PR TITLE
Initial PyTorch dataloaders from TFRecords

### LINKED ISSUE #455: Add pytorch Datasets loader to datasets
This requires using PyTorch DataLoader to read directly from the TFDS datasets. There may be some issues making this performant, as PyTorch uses very different methods of loading the data.
See this thread: https://discuss.pytorch.org/t/read-dataset-from-tfrecord-format/16409

Most of what I have found focuses on reading TFRecord files (which are protobufs), e.g., https://github.com/vahidk/tfrecord
but does not leverage the larger dataset_info.json and related txt/json files that would be necessary for parsing an entire dataset automatically.

### PR DESCRIPTION
**Pull request by seanpmorgan**
Closes #455 

Works for cifar10 and mnist. Some design decisions we'd need to make and then we can expand this to other datasets.

### CODE REVIEW

File: armory/data/datasets.py [hunk 1]
```diff
@@ -212,58 +206,73 @@ def _generator_from_tfds(
...
-    if shuffle_files:
-        ds = ds.shuffle(batch_size * 10, reshuffle_each_iteration=True)
-    if variable_length and batch_size > 1:
-        ds = ds.batch(1, drop_remainder=False)
+        if not shuffle_files:
+            raise ValueError("PyTorch DataLoaders from Armory are shuffled by default")
+        ds_name, ds_version = dataset_name.split(":")
+
+        # TODO: Make this configurable. Possibly a mapping from dataset_name???
+        ds = ImageTFRecordDataSet(ds_name, ds_version, split_type)
```
Comment (seanpmorgan) [inline armory/data/datasets.py:217]: It'd be nice if we could parse the relevant JSON files from each dataset and have a generic dataset loader here, but it may be easier to just map them to respective loaders.

File: armory/data/pytorch_loaders.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        return np.array(Image.open(BytesIO(bytes_img[0])))
+
+    def preprocess(self, x):
+        label, image = x
+        image = self.decode_image(image)
+
+        if image.ndim == 2:
+            image = np.expand_dims(image, axis=-1)
+
+        # TODO: permute the shape to channels first as is typical for PyTorch???
```
Comment (seanpmorgan) [inline armory/data/pytorch_loaders.py:44]: Need a decision on this

File: tests/test_pytorch/test_framework_dataset.py [hunk 3]
```diff
@@ -2,20 +2,47 @@
...
-            epochs=1,
-            batch_size=16,
-            dataset_dir=DATASET_DIR,
-            framework="pytorch",
-        )
+def test_pytorch_generator_cifar10():
+    batch_size = 16
+    dataset = datasets.cifar10(
+        split_type="train",
+        epochs=1,
```
Comment (seanpmorgan) [inline tests/test_pytorch/test_framework_dataset.py:17]: Need to verify how this works for multiple epochs as well

File: armory/data/pytorch_loaders.py (hunk 2 above)
Comment (davidslater) [reply armory/data/pytorch_loaders.py:44]: This is tricky, as we would then want to permute the inference time channels when using numpy mode in the attack. Hmmm.

File: armory/data/datasets.py [hunk 4]
```diff
@@ -212,58 +206,73 @@ def _generator_from_tfds(
...
+
+        generator = torch.utils.data.DataLoader(
+            ds, batch_size=batch_size, num_workers=2
+        )
+        return generator
```
Comment (davidslater) [inline armory/data/datasets.py:222]: It would be nice to make `num_workers` configurable.

File: armory/data/datasets.py (hunk 1 above)
Comment (davidslater) [reply armory/data/datasets.py:217]: For now, let's just hard map them in.

File: armory/data/datasets.py [hunk 5]
```diff
@@ -212,58 +206,73 @@ def _generator_from_tfds(
...
-        ds = ds.batch(1, drop_remainder=False)
+        if not shuffle_files:
+            raise ValueError("PyTorch DataLoaders from Armory are shuffled by default")
+        ds_name, ds_version = dataset_name.split(":")
+
+        # TODO: Make this configurable. Possibly a mapping from dataset_name???
+        ds = ImageTFRecordDataSet(ds_name, ds_version, split_type)
+
+        generator = torch.utils.data.DataLoader(
+            ds, batch_size=batch_size, num_workers=2
```
Comment (davidslater) [inline armory/data/datasets.py:220]: This needs to have a `worker_init_fn` defined to handle multi-process data loading. Currently, each worker will return duplicate batches.

See the second example in: https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset

Example:
```
dataset = datasets.cifar10(
         split_type="train",
         epochs=1,
         batch_size=batch_size,
         dataset_dir=DATASET_DIR,
         framework="pytorch",
     )
iter_ds = iter(dataset)
x1, y1 = next(iter_ds)
x2, y2 = next(iter_ds)
print((x1 == x2).all())
# tensor(True)
print((y1 == y2).all())
# tensor(True)
```

File: armory/data/pytorch_loaders.py (hunk 2 above)
Comment (davidslater) [reply armory/data/pytorch_loaders.py:44]: Let's go ahead and switch to channels first, as channels last is still experimental in pytorch.

File: tests/test_pytorch/test_framework_dataset.py (hunk 3 above)
Comment (davidslater) [reply tests/test_pytorch/test_framework_dataset.py:17]: It only generates a single epoch, regardless of what `epochs` is.

File: armory/data/datasets.py [hunk 6]
```diff
@@ -212,58 +206,73 @@ def _generator_from_tfds(
...
-        ds = ds.map(lambda_map)
-
-    ds = ds.repeat(epochs)
-    if shuffle_files:
-        ds = ds.shuffle(batch_size * 10, reshuffle_each_iteration=True)
-    if variable_length and batch_size > 1:
-        ds = ds.batch(1, drop_remainder=False)
+        if not shuffle_files:
+            raise ValueError("PyTorch DataLoaders from Armory are shuffled by default")
+        ds_name, ds_version = dataset_name.split(":")
```
Comment (davidslater) [inline armory/data/datasets.py:214]: We should probably document that this shuffling is due to dareblopy, not due to the defaults of PyTorch iter dataloaders.

File: armory/data/datasets.py (hunk 1 above)
Comment (davidslater) [reply armory/data/datasets.py:217]: If we are mapping, should we move all of the NotImplementedError throwing into here, instead of replicating across datasets?

File: tests/test_pytorch/test_framework_dataset.py (hunk 3 above)
Comment (davidslater) [reply tests/test_pytorch/test_framework_dataset.py:17]: You can call `iter(dataset)` again to start over, but the iteration is in the same order as the initial shuffle (it is not reshuffled).

File: armory/data/pytorch_loaders.py (hunk 2 above)
Comment (seanpmorgan) [reply armory/data/pytorch_loaders.py]: Per discussion we'll be leaving it in channels last format so that pytorch baseline models take the same inputs in numpy and torch

**Comment by ng390** [comment]
LGTM
